shell/browser/electron_browser_client.cc

```diff
@@ -1520,14 +1521,19 @@ bool ElectronBrowserClient::WillCreateURLLoaderFactory(
   if (header_client)
     header_client_receiver = header_client->InitWithNewPipeAndPassReceiver();
 
+  mojo::PendingRemote<network::mojom::URLLoaderFactory> proxy_factory_remote;
+
   auto* protocol_registry =
       ProtocolRegistry::FromBrowserContext(browser_context);
   new ProxyingURLLoaderFactory(
-      web_request.get(), protocol_registry->intercept_handlers(),
-      render_process_id, &next_id_, std::move(navigation_ui_data),
-      std::move(navigation_id), std::move(proxied_receiver),
+      web_request.get(), render_process_id, &next_id_,
+      std::move(navigation_ui_data), std::move(navigation_id),
+      proxy_factory_remote.InitWithNewPipeAndPassReceiver(),
       std::move(target_factory_remote), std::move(header_client_receiver),
       type);
+  new InterceptingURLLoaderFactory(protocol_registry->intercept_handlers(),
+                                   std::move(intercepted_receiver),
+                                   std::move(proxy_factory_remote));
```

Comment:
It seems odd to me that we need _two_ URLLoaderFactories. Can/should this be consolidated into one factory? Is there any performance benefit from doing so?

Reply:
> It seems odd to me that we need two URLLoaderFactories.

Having only one factory doing both intercepting and proxying made the change more awkward.

Since the factory is a `mojo::Remote`, the only entrypoint is `CreateLoaderAndStart`. By the time the `next()` is invoked from JS, the only reference to the proxying factory is as a `mojo::PendingRemote`, so the only thing you can really do with it is call back into `CreateLoaderAndStart`, which would just intercept the request again, unless there's some way to tell it not to.

> Can/should this be consolidated into one factory?

The original commit did that, you can [review that implementation](https://github.com/electron/electron/pull/26065/commits/690b711a092e52bd598f36573b46f771e18505a5) for a comparison. Landing #25393 made it become awkward which is why I abandoned it in favor of the latest implementation. Specifically what got awkward was I'd separated out the default loading implementation so avoid the issue mentioned above, but there's no way to access that when using the factory as a `mojo::Remote`, so I ended up using a raw pointer which felt wrong there.

> Is there any performance benefit from doing so?

Nothing intended there, I'd expect performance to be trivially worse just because of the extra calls involved.

Reply:
ok, makes sense. I think it's cheap + rare enough to not worry too much about the extra cost.

Reply:
Sounds good. Definitely a lesser of two evils kind of deal, but there is the benefit that having each factory only doing one thing makes it a bit easier to follow the code (IMO), and moves the proxying factory closer to `extensions::WebRequestProxyingURLLoaderFactory` so it should be a bit easier to keep them in sync now that the extra functionality is pulled out.

I'll work the other comments now that this is resolved.